### PR TITLE
Lower randint default dtype to the C++ API

### PR DESCRIPTION
**Pull request by peterbell10**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #81380
* #81379
* #81378
* __->__ #81410
* #81479

The default dtype for randint is currently handled with manual python
binding code, this moves it into the `native_functions.yaml` declaration 
for API consistency.

BC-breaking message:

Previously `torch::randint` would default to the global default data type, which is usually `float`. This changes the default to int64, in order to match the python API.

### CODE REVIEW

File: test/test_jit.py [hunk 1]
```diff
@@ -11244,14 +11244,12 @@ def test_rand():
         def randint():
             return torch.randint(0, 5, [1, 2])
 
-        # although the type should be int here, testing that the runtime dtype
-        # and shape analysis dtype is the same.
```
Comment (peterbell10) [inline test/test_jit.py:11248]: Does this need a deprecation cycle even though it's a bug?

File: torch/csrc/autograd/python_torch_functions_manual.cpp [hunk 2]
```diff
@@ -413,7 +413,7 @@ static PyObject* THPVariable_randint(
       auto size = r.intlist(1);
       auto generator = r.generator(2);
       // NOTE: r.scalartype(X) gives the default dtype if r.isNone(X)
-      auto dtype = r.scalartypeWithDefault(4, at::ScalarType::Long);
+      auto dtype = r.scalartypeOptional(4);
```
[the remark this answers is not in the document]
Comment (peterbell10) [reply torch/csrc/autograd/python_torch_functions_manual.cpp:416]: The current codegen always defaults to the default floating type, and never passes through `nullopt`. The next PR in the stack fixes that and removes the manual bindings for several factory functions.

File: aten/src/ATen/native/TensorFactories.cpp [hunk 3]
```diff
@@ -1157,11 +1158,12 @@ Tensor bartlett_window(int64_t window_length,
 Tensor bartlett_window(
     int64_t window_length,
     bool periodic,
-    c10::optional<ScalarType> dtype,
+    c10::optional<ScalarType> dtype_opt,
     c10::optional<Layout> layout,
     c10::optional<Device> device,
     c10::optional<bool> pin_memory) {
   // See [Note: hacky wrapper removal for TensorOptions]
+  ScalarType dtype = c10::value_or_else(dtype_opt, [] { return c10::get_default_dtype_as_scalartype(); });
```
[the remark this answers is not in the document]
Comment (peterbell10) [reply aten/src/ATen/native/TensorFactories.cpp:1166]: Actually looking at this again, this is in the wrong PR. The problem comes from calling `arange` with `dtype=c10::nullopt` which defaults to `kLong`. However, that doesn't relate to `randint`.

**Review by albanD — APPROVED** [review]
This is only BC-breaking for the c++ API right?
Worth doing in any case I think.

**Comment by peterbell10** [comment]
Yes, the only Python changed should be `torch.jit.script` now giving the correct default type to `torch.randint`.

**Comment by albanD** [comment]
Ok!
Feel free to merge!

**Pull request closed without merging**